docs/proposals/20210716-async-azure-resource-creation-deletion.md

```diff
@@ -0,0 +1,320 @@
+    - @devigned
+reviewers:
+    - TBD
+creation-date: 2021-07-16
+last-updated: 2021-07-26
+status: implementable
+see-also:
+    - https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/1181
+    - https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/1067
+---
+
+# Async Azure Resource Creation and Deletion
+
+## <a name='TableofContents'></a>Table of Contents
+
+<!-- vscode-markdown-toc -->
+* [Table of Contents](#TableofContents)
+* [Summary](#Summary)
+* [Motivation](#Motivation)
+	* [Goals](#Goals)
+	* [Non-Goals / Future Work](#Non-GoalsFutureWork)
+* [Proposal](#Proposal)
+	* [User Stories](#UserStories)
+		* [Story 1 - UX of creating an AzureCluster](#Story1-UXofcreatinganAzureCluster)
+		* [Story 2 - Scaling up a MachineDeployment](#Story2-ScalingupaMachineDeployment)
+		* [Story 3 - Deleting an individual Azure Machine Pool Machine](#Story3-DeletinganindividualAzureMachinePoolMachine)
+	* [Implementation Details/Notes/Constraints](#ImplementationDetailsNotesConstraints)
+	* [Proposed API Changes](#ProposedAPIChanges)
+	* [Proposed Controller Changes](#ProposedControllerChanges)
+		* [Context timeouts](#Contexttimeouts)
+		* [Service Reconcile](#ServiceReconcile)
+		* [Service Delete](#ServiceDelete)
+		* [AzureCluster Reconcile](#AzureClusterReconcile)
+	* [Proposed New Conditions](#ProposedNewConditions)
+	* [Open Questions](#OpenQuestions)
+		* [1. What should the timeout durations be?](#Whatshouldthetimeoutdurationsbe)
+* [Alternatives](#Alternatives)
+	* [Parallel reconciliation of Azure services](#ParallelreconciliationofAzureservices)
+		* [Pros](#Pros)
+		* [Cons](#Cons)
+		* [Conclusion](#Conclusion)
+* [Additional Details](#AdditionalDetails)
+	* [Test Plan](#TestPlan)
+* [Implementation History](#ImplementationHistory)
+
+<!-- vscode-markdown-toc-config
+	numbering=false
+	autoSave=false
+	/vscode-markdown-toc-config -->
+<!-- /vscode-markdown-toc -->
+
+## <a name='Summary'></a>Summary
+
+CAPZ reconcilers currently call Azure and wait for each operation before proceeding. We should create/update/delete Azure resources asynchronously, especially for operations that take a long time to complete, such as Virtual Machine creation and deletion.
+
+## <a name='Motivation'></a>Motivation
+
+Blocking on success is sometimes the right thing to do but most of the time, it's the equivalent of the UI freezing on an app because you have used the UI thread to fetch some data causing your user to wonder why and when the software will react. This proposal aims to make the reaction time of the CAPZ controller drastically faster and possibly, more resilient.
+
+### <a name='Goals'></a>Goals
+
+- Accelerate the feedback loop with the user so they can know that reconciliation is progressing without having to go check the Azure resources in the portal/CLI/etc.
+- Make the time for the controller to react to a change much faster
+- Improve the resiliency of the controller by making it more fault tolerant
```

Comment:
Is the additional fault tolerance because we don't have to fail a bunch of resource reconciliations if just one resource reconciliation fails?

Reply:
> we don't have to fail a bunch of resource reconciliations if just one resource reconciliation fails?

If one resource fails, we short circuit the entire loop and try again later. However, if one resource is stuck creating in Azure or is taking longer than normal to create/update/delete, we won't waste the controller loop on that resource by polling continuously.